runtime/jcl/common/jdk_internal_foreign_abi_UpcallStubs.cpp

```diff
@@ -0,0 +1,124 @@
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "j9protos.h"
+#include "jcl.h"
+#include "jclglob.h"
+#include "jclprots.h"
+#include "jcl_internal.h"
+#include "jni.h"
+
+extern "C" {
+
+#if JAVA_SPEC_VERSION >= 16
+
+#if defined(OSX) && defined(AARCH64)
+#include <pthread.h> // for pthread_jit_write_protect_np
+#endif
+
+/**
+ * A memory segment is associated with a native scope/session owned by a thread, in which case
+ * the memory segment belonging to the scope/session will be automatically released in OpenJDK
+ * by invoking this function when the native scope/session is terminated.
+ * See src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/UpcallStubs.java
+ * in OpenJDK for details.
+ *
+ * To adapt to the calling mechanism in our implementation, a hashtable is created to stored
+ * the metadata for each generated thunk, which means the allocated memories associated with
+ * the metadata and the thunk are released in this way to ensure we have sufficient memory
+ * to be allocated on the heap for thunk (especially on AIX only with 4KB virtual memory).
+ */
+jboolean JNICALL
+Java_jdk_internal_foreign_abi_UpcallStubs_freeUpcallStub0(JNIEnv *env, jobject receiver, jlong address)
+{
+	J9VMThread *currentThread = (J9VMThread *)env;
+	J9JavaVM *vm = currentThread->javaVM;
+	const J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+#if defined(AIXPPC) || (defined(S390) && defined(zOS))
+	/* The first element of the function pointer holds the thunk memory address,
+	 * as specified in vm/ap64/UpcallThunkGen.cpp
+	 */
+	UDATA *functionPtr = (UDATA *)address;
+	void *thunkAddr = (void *)functionPtr[0];
+#else /* defined(AIXPPC) || (defined(S390) && defined(zOS)) */
+	void *thunkAddr = (void *)(UDATA)address;
+#endif /* defined(AIXPPC) || (defined(S390) && defined(zOS)) */
+
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	omrthread_monitor_enter(vm->thunkHeapWrapperMutex);
+	if (NULL != thunkAddr) {
+		J9UpcallThunkHeapWrapper *thunkHeapWrapper = vm->thunkHeapWrapper;
+		J9HashTable *metaDataHashTable = thunkHeapWrapper->metaDataHashTable;
+		J9Heap *thunkHeap = thunkHeapWrapper->heap;
+
+		if (NULL != metaDataHashTable) {
+			J9UpcallMetaDataEntry metaDataEntry = {0};
+			metaDataEntry.thunkAddrValue = (UDATA)(uintptr_t)thunkAddr;
+			metaDataEntry.upcallMetaData = NULL;
+			J9UpcallMetaDataEntry * result = (J9UpcallMetaDataEntry *)hashTableFind(metaDataHashTable, &metaDataEntry);
+			if (NULL != result) {
+				J9UpcallMetaData *metaData = result->upcallMetaData;
+				if (NULL != metaData) {
```

Comment:
can this be NULL?

Reply:
Removed as the null check on the metadata is unnecessary given each generated thunk address should come with a unique non-null metadata.